sass-spec.js

```diff
@@ -0,0 +1,192 @@
+const { archiveFromStream } = require("node-hrx")
+const tap = require("tap")
+const { promises: fs, createReadStream, rmdir } = require("fs")
+const path = require("path")
+const yaml = require("js-yaml")
+const child_process = require("child_process")
+
+/** Returns whether an options.yml object has a todo for the given impl */
+function hasTodo(options, impl) {
+  if (!options[":todo"]) return false
+  return options[":todo"].some((item) => item.includes(impl))
+}
+
+function hasIgnore(options, impl) {
+  if (!options[":ignore_for"]) return false
+  return options[":ignore_for"].includes(impl)
+}
+
+async function writeArchive(basePath, item) {
+  const fullPath = path.resolve(basePath, item.path)
+  if (item.isDirectory()) {
+    // If a directory, make the directory and recurse
+    await fs.mkdir(fullPath)
+    for (const subitem of item) {
+      await writeArchive(basePath, item.contents[subitem])
+    }
+  } else {
+    // We're a file, so write to it
+    await fs.writeFile(fullPath, item.body, { encoding: "utf-8" })
+  }
+}
+
+/**
+ * Unarchives the given HRX archive into the filesystem
+ */
+async function unarchive(parentDir, archiveName) {
+  // make a directory for the archive in the given directory
+  const dirName = archiveName.replace(".hrx", "")
+  const dirPath = path.resolve(parentDir, dirName)
+
+  // Unarchive and read the contents
+  const archivePath = path.resolve(parentDir, archiveName)
+  const archive = await archiveFromStream(
+    createReadStream(archivePath, { encoding: "utf-8" })
+  )
+  await writeArchive(dirPath, archive)
+}
+
+/**
+ * Run through the given directory
+ * @param dir the directory to iterate over
+ * @param opts the options to run the thing with:
+ *    - impl -- the sass implementation to use
+ *    - todo -- whether todos should be run
+ * @param cb the callback to run on each directory
+ */
+async function iterateDir(dir, opts, cb) {
```

Comment:
It's probably premature right now, but I recommend using the Ruby [`Directory` class](https://github.com/sass/sass-spec/blob/master/lib/sass_spec/directory.rb) as inspiration for an API for accessing files transparently between HRX and the physical filesystem. It's one of the newer and cleaner additions to the existing spec runner, and it's been very helpful as a basis for more complex features like the `--interactive` flag.

Reply:
I have looked at it! I just wanted to start out with something simpler because it seemed like a pretty feature-rich class and I wasn't sure which features we still wanted to include or not.